### DISCUSSION TITLE
I created a master/slave Prisma proxy with PostgreSQL + NextJS

### DISCUSSION THREAD

**Discussion by zirlman** [prisma/prisma#20539] [Q&A]
### Question

Hello everyone,

The last 5 days I spent configuring a master-slave architecture using PostgreSQL, Prisma & NextJS with the app dir. 
Because this was a real pain in the a$$ since I haven't done this stuff before, I decided to share my Prisma client so other developers can save hours of mental rollercoaster. 

Before I show you the Prisma setup, here's how my DB is configured:
1 master PostgreSQL server in Docker 
2 slave servers in Docker
[haproxy](http://www.haproxy.org/) load balancer (also in docker)

I'm using docker images from [bitnami](https://hub.docker.com/r/bitnami/postgresql/) since they're very easy to set up. You just need to configure environment variables and you're ready to go. They have many docker-compose examples on their GitHub page, including a file for [master-slave replication](https://github.com/bitnami/containers/blob/main/bitnami/postgresql/docker-compose-replication.yml). 
Even though this looks very easy. I spent 2 days configuring it manually 🗡️  with some shell scripts until I stumbled across their image. 

Now let's get to the funny stuff.

This is what my `datasource` looks like:
```
datasource db {
  provider = "postgresql"
  url      = env("READ_DATABASE_URL")
  directUrl = env("WRITE_DATABASE_URL")
}
```

For the client, I've used Javascript [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) object. Basically what I've done is, I'm checking the function name that is called on my Prisma Proxy, to determine which client to use. If it's a `write`, `update` or `delete` operation I use the master db, else I use slave db. Here's what it looks like:
```ts
function isPrismaFunction(functionName: string, client: PrismaClient) {
  return typeof (client as any)[functionName] === "function";
}
function isPrismaProperty(propName: string, client: PrismaClient) {
  return (
    typeof (client as any)[propName] !== "undefined" &&
    typeof (client as any)[propName] !== "function"
  );
}

function isWriteOperation(functionName: string): boolean {
  return (
    functionName.startsWith("create") ||
    functionName.startsWith("update") ||
    functionName.startsWith("delete")
  );
}

export function getPrismaProxy(master: PrismaClient, slave: PrismaClient) {
  return new Proxy<PrismaClient>({} as typeof prismaMaster, {
    get: function (target, prop, receiver) {
      const functionName = String(prop);

      // Check if it's a valid Prisma function or property
      if (isPrismaFunction(functionName, master)) {
        return async (...args: any[]) => {
          const isWrite = isWriteOperation(functionName);
          const prisma = isWrite ? prismaMaster : prismaSlave;
          const result = await (prisma as any)[functionName](...args);
          return result;
        };
      } else if (isPrismaProperty(functionName, master)) {
        // If it's a property (not a function), return the corresponding property from the master client
        // @ts-ignore
        return prismaMaster[functionName];
      }

      // Property or function not found, return undefined
      return undefined;
    },
  });
}
```

And finally, to prevent constant reconnection attempts during hot-reload, I'm binding my Prisma object to global this:
```ts
const globalForPrisma = globalThis as unknown as {
  prisma: PrismaClient | undefined;
};

let prisma: PrismaClient;

if (process.env.NODE_ENV === "production") {
  prismaMaster = new PrismaClient({
    log: ["error","warn"],
    datasources: {
      db: {
        url: process.env.WRITE_DATABASE_URL,
      },
    },
  });
  prismaSlave = new PrismaClient({
    log: ["error", "warn"],
    datasources: {
      db: {
        url: process.env.READ_DATABASE_URL,
      },
    },
  });

  prisma = getPrismaProxy(prismaMaster, prismaSlave);
} else {
  if (!globalForPrisma.prisma) {
    prismaMaster = new PrismaClient({
      log: ["error", "warn"],
      datasources: {
        db: {
          url: process.env.WRITE_DATABASE_URL,
        },
      },
    });
    prismaSlave = new PrismaClient({
      log: ["error", "warn"],
      datasources: {
        db: {
          url: process.env.READ_DATABASE_URL,
        },
      },
    });

    globalForPrisma.prisma = getPrismaProxy(prismaMaster, prismaSlave);
  }

  prisma = globalForPrisma.prisma;
}

export { prisma };
```

What do you think of this approach? I'd love to hear your feedback. 

P.S. At the time of writing, I spend the last day debugging, only to figure out that Prisma v5.1.0 has some [bugs with enums](https://github.com/prisma/prisma/issues/20480), so please upgrade to v5.1.1 ASAP

**Comment by ludralph** [comment]
Hi @zirlman 👋 

Thank you for sharing the solution that worked for you. Your approach is quite interesting, and you've put a lot of thought into optimizing database operations for read and write queries. It provides a good balance between read and write operations, minimizes database connections, and offers a high level of control over the routing of queries to the appropriate database servers. However, Prisma's API could evolve, and it's important to ensure Prisma Proxy remains compatible. 

I do think the community will find this helpful.
